primitives/src/types/primitive_market_builder.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2024 Forecasting Technologies LTD.
+//
+// This file is part of Zeitgeist.
+//
+// Zeitgeist is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the
+// Free Software Foundation, either version 3 of the License, or (at
+// your option) any later version.
+//
+// Zeitgeist is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{
+    traits::MarketBuilder,
+    types::{
+        Deadlines, EarlyClose, Market, MarketBonds, MarketCreation, MarketDisputeMechanism,
+        MarketPeriod, MarketStatus, MarketType, OutcomeReport, Report, ScoringRule,
+    },
+};
+use alloc::vec::Vec;
+use sp_runtime::{DispatchError, Perbill};
+
+/// A sample market builder struct used for testing. No verification is done when calling `build()`;
+/// use at your own risk!
+///
+/// Fields are deliberately kept public to allow the straightforward construction of builder objects
+/// in spots where correctness isn't the primary concern.
+///
+/// # Generics
+///
+/// * `AI`: The account ID type.
+/// * `BA`: The balance type.
+/// * `BN`: The block number type.
+/// * `M`: The moment/time type.
+/// * `A`: The asset type.
+/// * `MI`: The market ID type.
+#[derive(Clone)]
+pub struct PrimitiveMarketBuilder<AI, BA, BN, M, A, MI> {
+    pub market_id: Option<MI>,
+    pub base_asset: A,
+    pub creator: AI,
+    pub creation: MarketCreation,
+    pub creator_fee: Perbill,
+    pub oracle: AI,
+    pub metadata: Vec<u8>,
+    pub market_type: MarketType,
+    pub period: MarketPeriod<BN, M>,
+    pub deadlines: Deadlines<BN>,
+    pub scoring_rule: ScoringRule,
+    pub status: MarketStatus,
+    pub report: Option<Report<AI, BN>>,
+    pub resolved_outcome: Option<OutcomeReport>,
+    pub dispute_mechanism: Option<MarketDisputeMechanism>,
+    pub bonds: MarketBonds<AI, BA>,
+    pub early_close: Option<EarlyClose<BN, M>>,
+}
+
+macro_rules! impl_builder_methods {
+    ($($field:ident: $type:ty),*) => {
+        $(
+            fn $field(&mut self, $field: $type) -> &mut Self {
+                self.$field = $field;
+                self
+            }
+        )*
+    }
+}
+
+impl<AI, BA, BN, M, A, MI> MarketBuilder<AI, BA, BN, M, A, MI>
+    for PrimitiveMarketBuilder<AI, BA, BN, M, A, MI>
+{
+    fn build(self) -> Result<Market<AI, BA, BN, M, A, MI>, DispatchError> {
+        Ok(Market {
+            market_id: self.market_id.unwrap(),
```

Comment:
That's a good point.

Reply:
[052b047](https://github.com/zeitgeistpm/zeitgeist/pull/1248/commits/052b0476ff25e9005040ab19848f70c42818d958)

Reply:
I've moved away from using two `MarketBuilder` structs. Instead, the `MarketBuilder` is now in market-commons and so is the trait implementation without panickers. Seems cleanest right now.

This will, of course, have to be changed again once we've gotten rid of market edits and want to move the validation logic from `prediction-markets/src/lib.rs` to `MarketBuilderTrait::build()`. In that case, we'll move the trait implementation to prediction-markets but leave the struct definition in market-commons. This means we'll have to add a dummy trait implementation to market-commons for testing. Seems fair to me.